src/explore-education-statistics-ckeditor/plugins/comments/commentsEditing.js

```diff
@@ -0,0 +1,351 @@
+import ToggleResolveCommentCommand from './toggleResolveCommentCommand';
+
+export default class CommentsEditing extends Plugin {
+  constructor(editor) {
+    super(editor);
+    this.config = editor.config.get('comments');
+    this.selectedComment = '';
+    this.removedMarkers = [];
+  }
+
+  init() {
+    if (!this.config) {
+      return;
+    }
+    const { editor } = this;
+    const viewDocument = editor.editing.view.document;
+
+    this.defineSchema();
+    this.defineConverters();
+
+    /**
+     * Set up commands
+     */
+    this.editor.commands.add('addComment', new AddCommentCommand(this.editor));
+    this.editor.commands.add(
+      'addCommentPlaceholder',
+      new AddCommentPlaceholderCommand(this.editor),
+    );
+    const selectCommentCommand = new SelectCommentCommand(this.editor);
+    this.editor.commands.add('selectComment', selectCommentCommand);
+    const removeCommentCommand = new RemoveCommentCommand(this.editor);
+    this.editor.commands.add('removeComment', removeCommentCommand);
+    this.editor.commands.add(
+      'resolveComment',
+      new ToggleResolveCommentCommand(this.editor),
+    );
+
+    editor.on('ready', () => {
+      /**
+       * Watch for undo operations we need to handle.
+       */
+      editor.commands.get('undo').on('execute', () => {
+        const docVersion = editor.model.document.version;
+        const lastOperation = editor.model.document.history.getOperation(
+          docVersion - 1,
+        );
+
+        if (lastOperation.name && lastOperation.name.startsWith('comment:')) {
+          if (
+            lastOperation.batch.operations[0].name &&
+            lastOperation.batch.operations[0].name.startsWith(
+              'resolvedcomment:',
+            )
+          ) {
+            this.config.undoRedoComment(
+              'undoResolveComment',
+              lastOperation.name,
+            );
+            return;
+          }
+
+          this.removedMarkers = this.removedMarkers.filter(
+            marker => marker !== lastOperation.name,
+          );
+          this.config.undoRedoComment('undoRemoveComment', lastOperation.name);
+          return;
+        }
+
+        if (
+          lastOperation.name &&
+          lastOperation.name === 'commentplaceholder' &&
+          lastOperation.newRange &&
+          !lastOperation.oldRange
+        ) {
+          this.config.undoRedoComment(
+            'undoAddComment',
+            lastOperation.batch.operations[0].name,
+          );
+          return;
+        }
+
+        if (
+          lastOperation.name &&
+          lastOperation.name.startsWith('resolvedcomment:')
+        ) {
+          this.config.undoRedoComment(
+            'undoUnresolveComment',
+            lastOperation.name,
+          );
+        }
+      });
+
+      /**
+       * Watch for redo operations we need to handle.
+       */
+      editor.commands.get('redo').on('execute', () => {
```

Comment:
Not sure the redo is acting as expected. Finding the following:

1. If I try and redo a comment (after undoing it), there's a significant lag time before it actually loads back in. See the following video: https://www.loom.com/share/2b340f5d6cc643b3be48fa07e4b42bc8

2. If I undo a comment, then if click absolutely anything else on the screen, I seem to lose the ability to redo.

There's probably a bunch of other weird things that can happen with this stuff as well. We'll definitely want Luke to test this thoroughly.

Reply:
I think 1 is fixed. Haven't looked at 2 yet.

Reply:
https://dfedigital.atlassian.net/browse/EES-2961 raised for 2.